staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -69,6 +285,247 @@ func TestSimpleQueue(t *testing.T) {
+	})
+}
+
+// TestDelayedAddDropsWhenActivelyQueued calls for delayed adds should drop the delayed add if the item is aleady actively queued.
+func TestDelayedAddDropsWhenActivelyQueued(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "foo"
+	second := "foo"
+
+	// add first entry into the active queue
+	q.Add(first)
+	q.AssertState(t, QueueState{
+		active: []interface{}{first},
+	})
+
+	// try to add the same item with a delay, should have dropped the waiting request.
+	q.AddAfter(second, 1*time.Second)
+	q.AssertState(t, QueueState{
+		active: []interface{}{first},
+	})
+
+	// try to add the same item with a delay, buy with PermitActiveAndWaiting ...
+	q.AddWithOptions(second, DelayingOptions{Duration: 1 * time.Second, PermitActiveAndWaiting: true})
+	q.AssertState(t, QueueState{
+		active:  []interface{}{first},
+		waiting: []interface{}{second},
+	})
+}
+
+// TestOptionsTakeLongerTakeExisting - these options give the client an ability to perform a conditional add but
+// only when the item isn't already queued.  This fuctionality also applies to immediate adds made by the AddWithOptions
+// method.
+func TestOptionsTakeLongerTakeExisting(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "foo"
+	second := "foo"
+	third := "bar"
+
+	// add an item into the waiting queue.
+	q.AddAfter(first, 2*time.Second)
+
+	// attempt to immediately add an item using TakeLonger option
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeLonger})
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// attempt to immediately add an item using TakeExisting option
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// attempt to add an item not queued should queue successfully...
+	q.AddWithOptions(third, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		active:  []interface{}{third},
+		waiting: []interface{}{first},
+	})
+
+	// step past the original delayed item and process it from the active queue
+	fakeClock.Step(5 * time.Second)
+	if err := waitForAdded(q, 2); err != nil {
+		t.Fatalf("the first item should have been queued")
+	}
+	q.AssertState(t, QueueState{
+		active: []interface{}{first, third},
+	})
+	// pop them both off the queue
+	item, _ := q.Get()
+	q.Done(item)
+	item, _ = q.Get()
+	q.Done(item)
+
+	// add the second item again with DropIfWaiting which should now succeed
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		active: []interface{}{second},
+	})
+}
+
+// TestDelayedAddInsertSchemes - tests the differnent scheduling behaviours available when the same item is already waiting.
+func TestDelayedAddInsertSchemes(t *testing.T) {
```

Comment:
- I think it would be useful to have this test like it is, but with other items mingled in, so that we get changing of the order of items in the heap. 
- would be also good to add some usage of PermitActiveAndWaiting
- also it might help to split this test into 2 - 3 smaller ones

Reply:
not having enough items in the tests will prevent us from discovering issues like this one https://github.com/kubernetes/kubernetes/pull/112328#discussion_r1082833386